torch/csrc/distributed/c10d/TCPStore.cpp

```diff
@@ -360,12 +360,36 @@ TCPStore::TCPStore(std::string host, const TCPStoreOptions& opts)
     std::this_thread::sleep_for(std::chrono::milliseconds(distrib(gen)));
   }
 
-  client_ = detail::TCPClient::connect(addr_, opts);
-  // TCP connection established
-  C10D_DEBUG("TCP client connected to host {}:{}", addr_.host, addr_.port);
+  // Try connecting several times -- if the server listen backlog is full it may
+  // fail on the first send in validate.
+  auto deadline = std::chrono::steady_clock::now() + opts.timeout;
+  auto retry = 0;
+  do {
+    try {
+      client_ = detail::TCPClient::connect(addr_, opts);
+      // TCP connection established
+      C10D_DEBUG("TCP client connected to host {}:{}", addr_.host, addr_.port);
+
+      // client's first query for validation
+      validate();
+
+      // success
+      break;
+    } catch (const std::exception& ex) {
+      if (deadline >= std::chrono::steady_clock::now()) {
+        throw;
+      }
 
-  // client's first query for validation
-  validate();
+      C10D_WARNING(
+          "TCP client failed to connect/validate to host {}:{} retrying (try={}): {}",
```

Comment:
Include timeout and/or retry in this message?

Perhaps in the exception thrown above too to indicate we're giving after that timeout period.

Reply:
added timeout and added another message when we timeout